Avoid coercing all SELECTs with the same column name to the same value

boblail opened the pull request:
### Overview of Issue

Rails had been mapping types to columns in a Result Set by indexing types on columns' names. When two values in a Result Set have the same column name, they are forced to share a type.

There are several scenarios where multiple values may share the same column name.

One involves 
1. One scenario (where two tables happen to have columns with the same name but the column's values differ) is explained in #36042. **This PR does not resolve the issue in #36042 but proved _part_ of a solution to _that_ issue.**

2. Another is when using functions in a `.select` or `.pluck`. In the absence of an Alias clause, Postgres names values with the last function called, so if two values in the same query happen to use `COALESCE`, they'll both be named `"coalesce"`.

   ```
   psql> SELECT COALESCE(NULL, 'four'), COALESCE(NULL, 4);
    coalesce | coalesce
   ----------+----------
    four     |        4
   (1 row)
   ```

   ```ruby
   ActiveRecord::Base.pluck(
     Arel.sql("COALESCE(NULL, 'four')"),
     Arel.sql("COALESCE(NULL, 4)")
   ) # => [[0, 4]]
   ```

I was able to reproduce this on Rails 4.2.11.1, 5.0.7.2, 5.1.7, 5.2.3, and master.

### Solution

This commit matches `types` with `columns` and `values` (in a row) by array position (which is an assumption already made in `PostgreSQL::DatabaseStatements#exec_query`). It adds `column_types` as a public method on `ActiveRecord::Result` to preserve the class's public API.

### Concerns

Does Rails expect client code to be constructing `ActiveRecord::Result` objects? Do we need to check whether its third argument is a `Hash`, give — maybe — a deprecation warning and convert it to an `Array`?

boblail committed:
Avoid coercing all SELECTs with the same column name to the same value

Fixes #36042

Rails had been mapping types to columns in a Result Set by indexing types on columns' names. When two values in a Result Set have the same column name, they are forced to share a type.

There are several scenarios where multiple values may share the same column name.

1. One (where two tables happen to have columns with the same name but the column's values differ) is outlined in #36042.

2. Another is when using functions in a `.select` or `.pluck`. In the absence of an Alias clause, Postgres names values with the last function called, so if two values in the same query happen to use `COALESCE`, they'll both be named `"coalesce"`.

   ```ruby
   ActiveRecord::Base.pluck(
     Arel.sql("COALESCE(NULL, 'four')"),
     Arel.sql("COALESCE(NULL, 4)")
   )
   ```

This commit matches `types` with `columns` and `values` (in a row) by array position (which is an assumption already made in `PostgreSQL::DatabaseStatements#exec_query`). It adds `column_types` as a public method on `ActiveRecord::Result` to preserve the class's public API.

Does Rails expect client code to be constructing `ActiveRecord::Result` objects? Do we need to check whether its third argument is a `Hash`, give — maybe — a deprecation warning and convert it to an `Array`?

st0012 commented:
Just some suggestions after quick review:
- I think the solution makes sense, great job 😄 
- Can you fix the codeclimate issue? So it doesn't look broken on the PR list page.
- Since you this PR fixes #36042, can you add some test cases similar to the ones in that issue?

boblail commented:
@st0012, rebased. I think the specs were failing b/c of a flaky test. Rebuilt and it looks OK now.

I added two test cases for this. The second — `test_support_different_types_for_columns_with_the_same_name_but_different_tables` is the same scenario as in #36042.

boblail commented:
@kamipo, could I get 👀 on this PR? I know you've got a strong familiarity with ActiveRecord and Postgres. I'm not sure if mentioning you is the right protocol — sorry if it's not 😬. (I'm not quite sure how to promote my PR and solicit a review.)

🙏 Thanks!

st0012 commented on activerecord/lib/active_record/result.rb:
```diff
@@ -110,13 +110,16 @@ def cast_values(type_overrides = {}) # :nodoc:
         end
       else
-        types = columns.map { |name| column_type(name, type_overrides) }
+        types = []
```
```ruby
        types = columns.map.with_index { |name, i| column_type(name, i, type_overrides) }
```

st0012 commented on activerecord/lib/active_record/result.rb:
```diff
@@ -34,13 +34,13 @@ module ActiveRecord
     include Enumerable
 
-    attr_reader :columns, :rows, :column_types
+    attr_reader :columns, :rows, :types
```
I don't see a strong reason to rename the interface/instance variable to `types`. This can break something that's not covered by tests, e.g. the `initialize_copy` is now broken because it only copies `@column_type` but not `@types`.

boblail replied:
🤔 Good callout re: `initialize_copy`! I'll audit uses of the instance variable in this class.

I wouldn't expect the _instance variable_ to be part of this classes interface, though, and I did implement `column_types` (below) to maintain the classes interface.

My concern with simply reusing `@column_types` for the new third argument to `initialize` is that the _type_ of that instance variable has changed (from a dictionary of types correlated to columns by name to an array of types correlated to columns by position).

boblail replied:
`initialize_copy` appears to be the only place where `@column_types` was referenced — I made that change! 👍 ✅

st0012 commented on activerecord/lib/active_record/result.rb:
```diff
@@ -34,13 +34,13 @@ module ActiveRecord
       @rows         = rows
       @hash_rows    = nil
-      @column_types = column_types
+      @types        = types
```
Can change this into something like

```ruby
@column_types = if column_types.present?
  Hash[@columns.zip(column_types)]
else
  {}
end
```

Then we won't need to make `column_types` a standalone method.

boblail replied:
We don't want to _store_ type data like this, though, because — and this is the bug — if one name occurs more than once in the `@columns` array, then when we index them in a hash, only one type will be associated with that name.

##### Example:
```ruby
Hash[%w[name name title].zip([Integer, Date, String])] #=> { "name" => Date, "title" => String }
```

boblail replied:
(Sorry — to complete my thought: ... so we _have_ to store the Array and the Hash can be derived from the Array.)

boblail replied:
I did end up making this change!

eugeneius commented on activerecord/lib/active_record/result.rb:
```diff
@@ -127,14 +129,22 @@ def cast_values(type_overrides = {}) # :nodoc:
+
+    def column_types
+      if @types.present?
+        Hash[@columns.zip(@types)]
```
This method is sometimes called inside a loop:

https://github.com/rails/rails/blob/a3f62cbbe76f7f75c8442d5bab13a8f1bdfd00a7/activerecord/lib/active_record/relation/calculations.rb#L373

The result should be memoised so that it will only be calculated once, as it was before this change.

eugeneius replied:
I wonder if it's worth roflscaling this a bit, similar to `hash_rows` below, since this implementation allocates more intermediate objects than before (a two-element array for each column, and an array to hold them all).

boblail replied:
Since @st0012 was suggesting it earlier, I roflscaled this by assigning `@column_types` in `initialize`

eugeneius commented on activerecord/lib/active_record/result.rb:
```diff
@@ -127,14 +129,22 @@ def cast_values(type_overrides = {}) # :nodoc:
+    end
+
+    def column_types
+      if @types.present?
```
Since `@types` is an array, you can use `any?` instead of `present?` here – or swap the branches and use `empty?`.

boblail replied:
👍 I made this change.

eugeneius commented on activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:
```diff
@@ -96,12 +96,12 @@ def execute(sql, name = nil)
-              types[fname] = get_oid_type(ftype, fmod, fname)
+              types << get_oid_type(ftype, fmod, fname)
             end
             build_result(columns: fields, rows: result.values, column_types: types)
```
I think this `column_types:` parameter should be renamed to `types:`.

The default value is now of the wrong type too:

https://github.com/rails/rails/blob/a3f62cbbe76f7f75c8442d5bab13a8f1bdfd00a7/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L746

kamipo commented:
Actually this doesn't fix #36042, I'm not excited to change the `cast_values`' behavior which is private API unless it does fix public API (e.g. `pluck` etc).

boblail committed:
fixup! renamed :column_types to :types in build_result

boblail committed:
fixup! add test for pluck that shows what public behavior this change fixes

boblail committed:
fixup! precalculated `@column_types`

boblail commented:
@kamipo, great callout! This does **not** fix #36042. We'd have to handle `type_overrides` differently [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation/calculations.rb#L202).

I added a test, though, to demonstrate a bug with `pluck` that this PR _does_ still fix: https://github.com/rails/rails/pull/36186/commits/6b57bf940745427fcd410f6329d55b921d04c32f

kamipo commented:
I know this issue affects only for PostgreSQL adapter (originally works for other MySQL, SQLite3, Oracle, and SQL Server adapters), I'm still not excited extra allocation for all cases to address such a minor case.

Like as a workaround for #36042, is it not enough to give a column alias?

```ruby
relation.pluck(Arel.sql("COALESCE(format, 'unspecified') a, COALESCE(nullable_status, 0) b"))
```

boblail commented:
@kamipo,

Yep! Using an alias is a totally viable workaround! (When I first made this PR, my team also went that route: https://github.com/boblail/pluck_map/pull/14)

So if you want to decline this PR, we **_do_** have an acceptable path forward.

I just think that the behavior is surprising — especially since you get different results from MySQL and Sqlite3. For the developer troubleshooting the issue, the root cause is not immediately obvious!!

Only a developer who has spelunked in Rails' source would be able to tell that Rails keys type information on name and Postgres returns more-generic names than the other two databases:

> Here's the output from each database in its REPL:
> 
> ##### SQLite
> 
> ```
> sqlite> select COALESCE(NULL, 'four'), COALESCE(NULL, 4);
> COALESCE(NULL, 'four')  COALESCE(NULL, 4)
> ----------------------  -----------------
> four                    4
> ```
> 
> ##### MySQL
> ```
> mysql> select COALESCE(NULL, 'four'), COALESCE(NULL, 4);
> +------------------------+-------------------+
> | COALESCE(NULL, 'four') | COALESCE(NULL, 4) |
> +------------------------+-------------------+
> | four                   |                 4 |
> +------------------------+-------------------+
> 1 row in set (0.00 sec)
> ```
> 
> ##### PostgreSQL
> ```
> boblail=# select COALESCE(NULL, 'four'), COALESCE(NULL, 4);
>  coalesce | coalesce
> ----------+----------
>  four     |        4
> (1 row)
> ```
> 

* * *

Do you think it would make sense for Rails to raise an exception if an `ActiveRecord::Result` uses the same name more than once???

OR — the extra allocation you're referring to — is that the Hash of `@column_types` in addition to the Array of `@types`? Would this PR be acceptable if we did that allocation only on-demand or could do without it?

(Thanks for the feedback!! 😄)

simi commented:
As @boblail mentioned (and I agree)

```ruby
ActiveRecord::Base.pluck(
  Arel.sql("COALESCE(NULL, 'four')"),
  Arel.sql("COALESCE(NULL, 4)")
) # => [[0, 4]]
```

is not expected result. Anyway I agree with @kamipo this is big change to handle this (probably minor) problem. It would be nice to at least mention this in documentation or raise on unsupported (conflicting) input to not surprise developers (and keep the Rails experience nice).
